taichi/program/kernel.cpp

```diff
@@ -266,6 +266,33 @@ void Kernel::LaunchContextBuilder::set_arg_external_array(
   ctx_->set_device_allocation(arg_id, is_device_allocation);
 }
 
+void Kernel::LaunchContextBuilder::set_arg_external_array_with_shape(
+    int arg_id,
+    uintptr_t ptr,
+    uint64 size,
+    const std::vector<int64> &shape,
+    bool is_device_allocation) {
```

Comment:
Actually if we are going to separate external array and ndarray, is_device_allocation is always false for external array. So we can safely set it to false and remove this arg

Reply:
Good, didn't know that this flag is specific to Ndarray. Then this function looks neat.